src/julia.jl

```diff
@@ -33,4 +33,139 @@ function multireshape(parent::AbstractArray, dimss::Array{Tuple{Vararg{Int}}})
+
+# Arguments
+* `ondiag`: should the diagonal be computed, i.e. should the function be run for an element and itself)
+* `oneway`: should only `f(x,y)` be calculated when `f(x,y) == f(y,x)`?.
+"""
+function pairwise(f::Function, a, output::Symbol; oneway = true, ondiag = false)
+    in(output, [:vec, :vector, :Vector]) || throw(ArgumentError("Only vector and matrix output defined"))
+    n = length(a)
+    outlength = floor(Int, n * (n - 1) / (2*oneway) + n*Int(ondiag))
+    f = tryfunc(f, a[1], a[1])
+    first = f(a[1], a[1])
+    r = Vector{typeof(first)}(outlength)
+    pairwise!(f, a, r)
+    r
+end
+
+
+"""
+    pairwise!(f::Function, a, r)
+
+Apply the function `f` to all pairwise combinations of elements from `a` and
+storing the results in `r`. Setting keyword `symmetric = true` will only calculate
+each variable combination once for symmetric functions. This is useful if `f` is
+expensive.
+"""
+function pairwise!(f::Function, a, r::AbstractMatrix; symmetric = false)
+    f = tryfunc(f, a[1], a[1])
+    n = length(a)
+    size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r ($(size(r)), should be $((n, n)))"))
+    if symmetric
+        pairwise_symmetric!(f, a, r, n)
+    else
+        pairwise_default!(f, a, r, n)
+    end
+end
+
+function pairwise_default!(f, a, r, n)
+    @inbounds for j = 1:n
+        aj = a[j]
+        for i = 1:n
+            r[i,j] = f(a[i], aj)
+        end
+    end
+end
+
+function pairwise_symmetric!(f, a, r, n)
+    @inbounds for j = 1:n
+        aj = a[j]
+        for i = j:n
+            tmp = f(a[i], aj)
+            r[i,j] = tmp
+            r[j,i] = tmp
+        end
+    end
+end
+
+"""
+    pairwise!(f::Function, a, r::AbstractVector; kwargs...)
+
+Apply the function `f` to all pairwise combinations of elements from `a`, and
+store the results in `r`
+
+# Arguments
+* `ondiag`: should the diagonal be computed, i.e. should the function be run for an element and itself)
+* `oneway`: should only `f(x,y)` be calculated when `f(x,y) == f(y,x)`?.
+"""
+function pairwise!(f::Function, a, r::AbstractVector; oneway = true, ondiag = false)
+    f = tryfunc(f, a[1], a[1])
+    n = length(a)
+    outlength = floor(Int, n * (n - 1) / (2*oneway) + n*Int(ondiag))
+    length(r) == outlength || throw(DimensionMismatch("Incorrect size of r ($(length(r)), should be $outlength"))
+    idx = 0
+    @inbounds for j = 1 : n
+        aj = a[j]
+        for i = (oneway ? (j:n) : (1:n))
+            if i != j || ondiag
+                r[idx += 1] = f(a[i], aj)
+            end
+        end
+    end
+    r
+end
+
+
+
+# makes it possible to apply both functions that are defined for two arguments
+# (e.g. `distance(x, y)`) and for a vector with two elements (e.g. `sum([x, y])`)
+# I worry about abusing the try-catch syntax here, I have jumped through some
+# hoops to ensure error handling is sensible
+function tryfunc(f::Function, x, y)
```

Comment:
this style is very non-idiomatic and is a performance trap.  There's no particular reason to support both, since a suer can always do `pairwise((x,y) -> f([x,y]), ...)` if they need a function that expects an array.